### ISSUE TITLE
cursor maxTooltipDistance 0 does not work when a series has null values

### ISSUE THREAD

**Issue by tobiu** [amcharts/amcharts4#2314]
chart config (inside a webworker):

https://github.com/neomjs/neo/blob/dev/apps/covid/view/country/LineChartComponent.mjs

main thread:

https://github.com/neomjs/neo/blob/dev/src/main/AmCharts.mjs

i was following this one:

https://www.amcharts.com/docs/v4/tutorials/multi-series-shared-tooltip-with-colored-bullets/

> Limiting number of series tooltips via maxTooltipDistance

so i was using

```
cursor: {
                maxTooltipDistance: 0
            }
```

![Screenshot 2020-04-04 at 17 28 29](https://user-images.githubusercontent.com/1177434/78454840-db080b00-769a-11ea-9121-65b5ba8d3b91.png)

![Screenshot 2020-04-04 at 17 28 21](https://user-images.githubusercontent.com/1177434/78454843-de9b9200-769a-11ea-9e08-a531ec0e9d3e.png)

live demo:
https://neomjs.github.io/pages/node_modules/neo.mjs/dist/production/apps/covid/index.html#mainview=table&country=USA

**Comment by martynasma** [comment]
Since your live demo does not seem to be working, I tried it on a simpler setup. The `maxTooltipDistance` works fine even if one of the series has `null` instead of a value.

We'll need some way to reproduce this too look further.

**Comment by tobiu** [comment]
now you scared me. double-checked and the live demo is working.

make sure to open the link in an desktop env, best using chrome.

the data changed (the api cut off older dates), so a good new use case to reproduce the double tips is this one:

https://neomjs.github.io/pages/node_modules/neo.mjs/dist/production/apps/covid/index.html#mainview=table&country=New%20Zealand

![Screenshot 2020-04-05 at 16 10 20](https://user-images.githubusercontent.com/1177434/78500644-fab13900-7757-11ea-89a6-70513697e26c.png)

**Comment by martynasma** [comment]
Yup working fine now.

Can you point me to NZ data set so I can set up test case locally for debugging?

**Comment by martynasma** [comment]
BTW, have you tried setting `maxTooltipDistance = -1`?

**Comment by tobiu** [comment]
the new zealand data looks like:
```
const data = [
    {date: "2020-03-05T23:00:00.000Z", cases: 4, deaths: null, recovered: null, active: 4},
    {date: "2020-03-06T23:00:00.000Z", cases: 5, deaths: null, recovered: null, active: 5},
    {date: "2020-03-07T23:00:00.000Z", cases: 5, deaths: null, recovered: null, active: 5},
    {date: "2020-03-08T23:00:00.000Z", cases: 5, deaths: null, recovered: null, active: 5},
    {date: "2020-03-09T23:00:00.000Z", cases: 5, deaths: null, recovered: null, active: 5},
    {date: "2020-03-10T23:00:00.000Z", cases: 5, deaths: null, recovered: null, active: 5},
    {date: "2020-03-11T23:00:00.000Z", cases: 5, deaths: null, recovered: null, active: 5},
    {date: "2020-03-12T23:00:00.000Z", cases: 5, deaths: null, recovered: null, active: 5},
    {date: "2020-03-13T23:00:00.000Z", cases: 6, deaths: null, recovered: null, active: 6},
    {date: "2020-03-14T23:00:00.000Z", cases: 8, deaths: null, recovered: null, active: 8},
    {date: "2020-03-15T23:00:00.000Z", cases: 8, deaths: null, recovered: null, active: 8},
    {date: "2020-03-16T23:00:00.000Z", cases: 12, deaths: null, recovered: null, active: 12},
    {date: "2020-03-17T23:00:00.000Z", cases: 20, deaths: null, recovered: null, active: 20},
    {date: "2020-03-18T23:00:00.000Z", cases: 28, deaths: null, recovered: null, active: 28},
    {date: "2020-03-19T23:00:00.000Z", cases: 39, deaths: null, recovered: null, active: 39},
    {date: "2020-03-20T23:00:00.000Z", cases: 52, deaths: null, recovered: null, active: 52},
    {date: "2020-03-21T23:00:00.000Z", cases: 102, deaths: null, recovered: null, active: 102},
    {date: "2020-03-22T23:00:00.000Z", cases: 102, deaths: null, recovered: null, active: 102},
    {date: "2020-03-23T23:00:00.000Z", cases: 155, deaths: null, recovered: 12, active: 143},
    {date: "2020-03-24T23:00:00.000Z", cases: 205, deaths: null, recovered: 22, active: 183},
    {date: "2020-03-25T23:00:00.000Z", cases: 283, deaths: null, recovered: 27, active: 256},
    {date: "2020-03-26T23:00:00.000Z", cases: 368, deaths: null, recovered: 37, active: 331},
    {date: "2020-03-27T23:00:00.000Z", cases: 451, deaths: null, recovered: 50, active: 401},
    {date: "2020-03-28T23:00:00.000Z", cases: 514, deaths: 1, recovered: 56, active: 457},
    {date: "2020-03-29T22:00:00.000Z", cases: 589, deaths: 1, recovered: 63, active: 525},
    {date: "2020-03-30T22:00:00.000Z", cases: 647, deaths: 1, recovered: 74, active: 572},
    {date: "2020-03-31T22:00:00.000Z", cases: 708, deaths: 1, recovered: 83, active: 624},
    {date: "2020-04-01T22:00:00.000Z", cases: 797, deaths: 1, recovered: 92, active: 704},
    {date: "2020-04-02T22:00:00.000Z", cases: 868, deaths: 1, recovered: 103, active: 764},
    {date: "2020-04-03T22:00:00.000Z", cases: 950, deaths: 1, recovered: 127, active: 822},
    {date: 1586102912950, active: 882, cases: 1039, deaths: 1, recovered: 156}
];
```

just noticed that the "today" entry has a different date format, will fix this now.

i will also try -1 and give you a heads up.

best regards
tobias

**Comment by tobiu** [comment]
the date format did not matter, `maxTooltipDistance: -1` seems to do the trick.
i will do a fresh deployment for the online demos now.

**Comment by tobiu** [comment]
deployment done, looks fine inside the dist/production environment as well. thanks for the heads up! feel free to close this ticket (maybe add the -1 to the guide).

you might get a couple of new users in case the neo framework keeps gaining traction.

off topic [way ahead of the ticket scope]: at some point we could chat about improving the json syntax if you like to. since the config objects get send from a worker to the main thread (post messages) it is impossible to add functions or chart related variables into it (eval() is not an option). the path based callMethod and setProperty hacks work fine for now though.

**Comment by martynasma** [comment]
Added note about `-1` here:
https://www.amcharts.com/docs/v4/concepts/chart-cursor/#Limiting_number_of_series_tooltips

Classref already had it noted:
https://www.amcharts.com/docs/v4/reference/xycursor/#maxTooltipDistance_property

**Comment by tobiu** [comment]
i actually encountered some edge cases where i get 2 tips with using -1.
happens rarely though.

**Comment by martynasma** [comment]
If you can create some scaled down chart we can use for testing that shows the issue, we can look into it.

**Comment by tobiu** [comment]
good morning! i found an reproducible edge case.
again: it works a lot better with -1, so low prio item (in my opinion).

in the chart for the Philippines, hovering the green line at the bottom (x-axis) creates 2 tips.

![Screenshot 2020-04-06 at 10 16 39](https://user-images.githubusercontent.com/1177434/78537584-fe909a00-77ef-11ea-9857-2e5dc98e10ab.png)

![Screenshot 2020-04-06 at 10 15 27](https://user-images.githubusercontent.com/1177434/78537598-051f1180-77f0-11ea-8d64-f1574076899f.png)

![Screenshot 2020-04-06 at 10 16 03](https://user-images.githubusercontent.com/1177434/78537604-08b29880-77f0-11ea-976c-6fe016de1d2d.png)

**Comment by tobiu** [comment]
```
const data = [
    {date: "2020-01-29T23:00:00.000Z", cases: 1, deaths: null, recovered: null, active: 1},
    {date: "2020-01-30T23:00:00.000Z", cases: 1, deaths: null, recovered: null, active: 1},
    {date: "2020-01-31T23:00:00.000Z", cases: 1, deaths: null, recovered: null, active: 1},
    {date: "2020-02-01T23:00:00.000Z", cases: 2, deaths: 1, recovered: null, active: 1},
    {date: "2020-02-02T23:00:00.000Z", cases: 2, deaths: 1, recovered: null, active: 1},
    {date: "2020-02-03T23:00:00.000Z", cases: 2, deaths: 1, recovered: null, active: 1},
    {date: "2020-02-04T23:00:00.000Z", cases: 2, deaths: 1, recovered: null, active: 1},
    {date: "2020-02-05T23:00:00.000Z", cases: 2, deaths: 1, recovered: null, active: 1},
    {date: "2020-02-06T23:00:00.000Z", cases: 3, deaths: 1, recovered: null, active: 2},
    {date: "2020-02-07T23:00:00.000Z", cases: 3, deaths: 1, recovered: null, active: 2},
    {date: "2020-02-08T23:00:00.000Z", cases: 3, deaths: 1, recovered: null, active: 2},
    {date: "2020-02-09T23:00:00.000Z", cases: 3, deaths: 1, recovered: null, active: 2},
    {date: "2020-02-10T23:00:00.000Z", cases: 3, deaths: 1, recovered: null, active: 2},
    {date: "2020-02-11T23:00:00.000Z", cases: 3, deaths: 1, recovered: 1, active: 1},
    {date: "2020-02-12T23:00:00.000Z", cases: 3, deaths: 1, recovered: 1, active: 1},
    {date: "2020-02-13T23:00:00.000Z", cases: 3, deaths: 1, recovered: 1, active: 1},
    {date: "2020-02-14T23:00:00.000Z", cases: 3, deaths: 1, recovered: 1, active: 1},
    {date: "2020-02-15T23:00:00.000Z", cases: 3, deaths: 1, recovered: 1, active: 1},
    {date: "2020-02-16T23:00:00.000Z", cases: 3, deaths: 1, recovered: 1, active: 1},
    {date: "2020-02-17T23:00:00.000Z", cases: 3, deaths: 1, recovered: 1, active: 1},
    {date: "2020-02-18T23:00:00.000Z", cases: 3, deaths: 1, recovered: 1, active: 1},
    {date: "2020-02-19T23:00:00.000Z", cases: 3, deaths: 1, recovered: 1, active: 1},
    {date: "2020-02-20T23:00:00.000Z", cases: 3, deaths: 1, recovered: 1, active: 1},
    {date: "2020-02-21T23:00:00.000Z", cases: 3, deaths: 1, recovered: 1, active: 1},
    {date: "2020-02-22T23:00:00.000Z", cases: 3, deaths: 1, recovered: 1, active: 1},
    {date: "2020-02-23T23:00:00.000Z", cases: 3, deaths: 1, recovered: 1, active: 1},
    {date: "2020-02-24T23:00:00.000Z", cases: 3, deaths: 1, recovered: 1, active: 1},
    {date: "2020-02-25T23:00:00.000Z", cases: 3, deaths: 1, recovered: 1, active: 1},
    {date: "2020-02-26T23:00:00.000Z", cases: 3, deaths: 1, recovered: 1, active: 1},
    {date: "2020-02-27T23:00:00.000Z", cases: 3, deaths: 1, recovered: 1, active: 1},
    {date: "2020-02-28T23:00:00.000Z", cases: 3, deaths: 1, recovered: 1, active: 1},
    {date: "2020-02-29T23:00:00.000Z", cases: 3, deaths: 1, recovered: 1, active: 1},
    {date: "2020-03-01T23:00:00.000Z", cases: 3, deaths: 1, recovered: 1, active: 1},
    {date: "2020-03-02T23:00:00.000Z", cases: 3, deaths: 1, recovered: 1, active: 1},
    {date: "2020-03-03T23:00:00.000Z", cases: 3, deaths: 1, recovered: 1, active: 1},
    {date: "2020-03-04T23:00:00.000Z", cases: 3, deaths: 1, recovered: 1, active: 1},
    {date: "2020-03-05T23:00:00.000Z", cases: 5, deaths: 1, recovered: 1, active: 3},
    {date: "2020-03-06T23:00:00.000Z", cases: 6, deaths: 1, recovered: 1, active: 4},
    {date: "2020-03-07T23:00:00.000Z", cases: 10, deaths: 1, recovered: 1, active: 8},
    {date: "2020-03-08T23:00:00.000Z", cases: 20, deaths: 1, recovered: 1, active: 18},
    {date: "2020-03-09T23:00:00.000Z", cases: 33, deaths: 1, recovered: 2, active: 30},
    {date: "2020-03-10T23:00:00.000Z", cases: 49, deaths: 1, recovered: 2, active: 46},
    {date: "2020-03-11T23:00:00.000Z", cases: 52, deaths: 2, recovered: 2, active: 48},
    {date: "2020-03-12T23:00:00.000Z", cases: 64, deaths: 5, recovered: 2, active: 57},
    {date: "2020-03-13T23:00:00.000Z", cases: 111, deaths: 8, recovered: 2, active: 101},
    {date: "2020-03-14T23:00:00.000Z", cases: 140, deaths: 11, recovered: 2, active: 127},
    {date: "2020-03-15T23:00:00.000Z", cases: 142, deaths: 12, recovered: 2, active: 128},
    {date: "2020-03-16T23:00:00.000Z", cases: 187, deaths: 12, recovered: 5, active: 170},
    {date: "2020-03-17T23:00:00.000Z", cases: 202, deaths: 19, recovered: 5, active: 178},
    {date: "2020-03-18T23:00:00.000Z", cases: 217, deaths: 17, recovered: 8, active: 192},
    {date: "2020-03-19T23:00:00.000Z", cases: 230, deaths: 18, recovered: 8, active: 204},
    {date: "2020-03-20T23:00:00.000Z", cases: 307, deaths: 19, recovered: 13, active: 275},
    {date: "2020-03-21T23:00:00.000Z", cases: 380, deaths: 25, recovered: 17, active: 338},
    {date: "2020-03-22T23:00:00.000Z", cases: 462, deaths: 33, recovered: 17, active: 412},
    {date: "2020-03-23T23:00:00.000Z", cases: 552, deaths: 35, recovered: 20, active: 497},
    {date: "2020-03-24T23:00:00.000Z", cases: 636, deaths: 38, recovered: 26, active: 572},
    {date: "2020-03-25T23:00:00.000Z", cases: 707, deaths: 45, recovered: 28, active: 634},
    {date: "2020-03-26T23:00:00.000Z", cases: 803, deaths: 54, recovered: 31, active: 718},
    {date: "2020-03-27T23:00:00.000Z", cases: 1075, deaths: 68, recovered: 35, active: 972},
    {date: "2020-03-28T23:00:00.000Z", cases: 1418, deaths: 71, recovered: 42, active: 1305},
    {date: "2020-03-29T22:00:00.000Z", cases: 1546, deaths: 78, recovered: 42, active: 1426},
    {date: "2020-03-30T22:00:00.000Z", cases: 2084, deaths: 88, recovered: 49, active: 1947},
    {date: "2020-03-31T22:00:00.000Z", cases: 2311, deaths: 96, recovered: 50, active: 2165},
    {date: "2020-04-01T22:00:00.000Z", cases: 2633, deaths: 107, recovered: 51, active: 2475},
    {date: "2020-04-02T22:00:00.000Z", cases: 3018, deaths: 136, recovered: 52, active: 2830},
    {date: "2020-04-03T22:00:00.000Z", cases: 3094, deaths: 144, recovered: 57, active: 2893},
    {date: "2020-04-04T22:00:00.000Z", cases: 3246, deaths: 152, recovered: 64, active: 3030},
    {date: "2020-04-06T08:30:16.273Z", active: 3198, cases: 3414, deaths: 152, recovered: 64}
];
```

**Comment by martynasma** [comment]
Thanks. Was able to finally reproduce it:

https://codepen.io/team/amcharts/pen/1e9d7f9b3c9479c6e2ac9de62154bfa6?editors=0010

Looking into it.

**Comment by martynasma** [comment]
Fixed in [4.9.14](https://github.com/amcharts/amcharts4/releases/tag/4.9.14).

## [4.9.14] - 2020-04-13

### Added
- Global options setting added `am4core.options.animationsEnabled` (default `true`). Set it to `false` to **disable all animations on all charts**. Can be used to toggle animations on and off dynamically.
- `SliceGrouper` plugin: new setting `limit` added. If set to a number, will ignore `threshold` but will show only first X slices and will group the rest.
- New PDF format option for `Export`: `align`. Available options: `"left"` (default), `"center"`, and `"right"`.
- New `ValueAxis` property: `adjustLabelPrecision` (default: `true`). `true` means that **all** labels on `ValueAxis` (except zero) will maintain the same number of decimals, e.g. `1.0`, `1.5`, `2.0` versus `1`, `1.5`, `2`. Enabling it will ignore number of active decimals set via `numberFormat`.
- `DateFormatter` is now exposing list of various named time units: `months`, `monthsShort`, `weekdays`, and `weekdaysShort`. You can now resolve numeric value to an English name (e.g. `chart.dateFormatter.weekdaysShort[2]`) or to your active locale (`chart.dateFormatter.language.translate(chart.dateFormatter.weekdaysShort[1])`).

### Changed
- New property `adjustLabelPrecision` hsa a default of `true`, which means that all labels will now have same number of decimals on `ValueAxis`. Set `adjustLabelPrecision = false`, to disable and revert behavior to how it was before this update (`1`, `1.5`, `2`).
- `am4core.options.viewportTarget` now accepts array of references if you have multiple scrollable containers holding charts and have `onlyShowOnViewport` enabled.
- Default `DateAxis.periodChangeDateFormats` for `"hour"` changed from `"_date_hour"` to `"_date_day"` so that the format would differ for midnight (hour 00:00).

### Fixed
- Parsing short month names (`"MMM"`) from string-based dates was failing on May.
- When exporting chart to PDF, on narrow charts, sometimes the whole chart would go to the next page.
- Having `null` values in data was causing `maxTooltipDistance == -1` to show multiple tooltips instead of one.
- It was impossible to have individual `innerRadius` for `AxisRendererRadial` and `AxisRendererCircular` - the `RadarChart`'s `innerRadius` was overriding them.
- Removing axis ranges from an axis could result in a JS error.
- In some cases, when `DateAxis` properties `startLocation` and `endLocation` were not defaults (0 and 1), the chart might flicker from one grouped data period to another continously. It was only happening with some particular `groupCount` values and data item counts, though.

[Full change log](https://github.com/amcharts/amcharts4/blob/master/dist/es2015/CHANGELOG.md).

Make sure you clear your browser cache after upgrading. And feel free to contact us again if you are still experiencing this issue.
